baseband/vdif/base.py

```diff
@@ -308,15 +311,31 @@ def __init__(self, fh_raw, thread_ids=None, sample_rate=None,
         # Now also read the first frameset, since we need to know how many
         # threads there are, and what the frameset size is. For this purpose,
         # pre-attach the file handle (it will just get reset anyway).
-        # TODO: make this a bit less ugly. Can we not just super first?
+        # TODO: make this a bit less ugly?  Maybe don't read frameset twice?
         fh_raw.seek(0)
         self.fh_raw = fh_raw
-        self._frameset = self.read_frameset(thread_ids)
-        if thread_ids is None:
-            thread_ids = [fr['thread_id'] for fr in self._frameset.frames]
+        self._frameset = self.read_frameset()
+        thread_ids = [fr['thread_id'] for fr in self._frameset.frames]
         self._framesetsize = fh_raw.tell()
-        super(VDIFStreamReader, self).__init__(fh_raw, header, thread_ids,
+        super(VDIFStreamReader, self).__init__(fh_raw, header, subset,
+                                               len(self._frameset.frames),
                                                sample_rate, squeeze)
+        # Set _thread_ids.  If subsetting, decode first frameset again.
+        if self.subset:
+            # Squeeze in case subset[0] uses broadcasting.
+            subset_0 = (self.subset[0].squeeze()
```

Comment:
I think this can be simplified by just doing
```
thread_ids = np.array(thread_ids)[subset[0]].squeeze()
```
I'll do it in a quick separate commit.

Reply:
Nope, that squeezes `np.array([1])` into `np.array(1)`.